### PR TITLE
Move RNN implementations to C++

### PR DESCRIPTION
**Pull request by apaszke**
This is the first of two changes that are supposed to improve how we handle RNNs in the JIT. They still get traced as `PythonOp`s, but now it will be much easier to actually expose them to the JIT as e.g. `aten::lstm`, and ignore the Python interpreter entirely. This needs some symbolic adjustments that will be part of a second PR.

Even when we fix symbolics, there will still be a bit of a problem with statefulness of the cuDNN API (we need a mutable cache for the dropout state, but our IR has no way of representing that).

@zdevito @ezyang

### CODE REVIEW

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+
+namespace {
+
+template<typename T>
+using pair_of = std::pair<T, T>;
+
+template<typename T>
+using tpair_of = std::tuple<T, T>;
+
+struct PackedSequence {
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: The old `torch.nn.utils.rnn.PackedSequence` had docs, this should too!

File: aten/src/ATen/native/RNN.cpp [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+}
+tpair_of<Tensor> hidden_slice(const tpair_of<Tensor>& t, int64_t start, int64_t end) {
+  return std::make_tuple(hidden_slice(std::get<0>(t), start, end),
+                         hidden_slice(std::get<1>(t), start, end));
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// CELL IMPLEMENTATIONS
+//
+// Cell is a basic component of an RNN, representing a single application of the
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: nit: Move this comment below to Cell? You read it, and then you read `linear` and you're like, "This isn't a cell..."

File: aten/src/ATen/native/RNN.cpp [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+// recurrent function. You can think of it as a function of signature
+//
+// (Tensor input, hidden_type hidden, CellParams) -> hidden_type
+//
+// which means that it consumes an input tensor, and updates the previous hidden state.
+// It's a struct only because functional programming in C++ is a pain, and it's easier
+// to pass around "vtable pointers" than actual function pointers.
+
+Tensor linear(const Tensor& input, const Tensor& weight, /* optional */ const Tensor& bias={}) {
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: Should this be a real ATen native function?

File: aten/src/ATen/native/RNN.cpp [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+  return output;
+}
+
+template<typename hidden_type_tmpl>
+struct Cell {
+  using hidden_type = hidden_type_tmpl;
+  virtual hidden_type operator()(const Tensor& input, const hidden_type& hidden, const CellParams& params) const = 0;
+};
+
+template<Tensor (*nonlinearity)(const Tensor&)>
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: nit: Give a name for the type of a nonlinearity function, it's easier to understand than the anonymous syntax for function pointer (blugh).

File: aten/src/ATen/native/RNN.cpp [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+    auto unstacked_output = (*this)(inputs.unbind(0), input_hidden, params);
+    return {at::stack(unstacked_output.outputs, 0), unstacked_output.final_hidden};
+  }
+
+  Cell<hidden_type>& cell_;
+};
+
+template<typename dir_hidden_type>
+struct FullBidirectionalLayer : Layer<Tensor, pair_of<dir_hidden_type>, pair_of<CellParams>> {
+  using hidden_type = pair_of<dir_hidden_type>;
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: Oh how I wish C++ had dependent types, so you wouldn't have to pass in the `hidden_type` in the template parameters of the parent class :)

**Comment by ezyang** [comment]
Haven't finished yet. If you want to merge this AOT and fix up residual comments later I'm OK with that.

**Comment by apaszke** [comment]
Thanks for a quick review, but I'll wait. I can find unrelated tasks to work on in the meantime.

File: aten/src/ATen/native/RNN.cpp [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+};
+
+template<typename io_type, typename hidden_type, typename param_type>
+struct Layer {
+  using output_type = LayerOutput<io_type, hidden_type>;
+  virtual output_type operator()(const io_type& input, const hidden_type& input_hidden, const param_type& params) const = 0;
+};
+
+template<typename hidden_type>
+struct FullLayer : Layer<Tensor, hidden_type, CellParams> {
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: Is "full" the standard nomenclature for this sort of layer?

File: aten/src/ATen/native/RNN.cpp [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+    auto rev_step_inputs = reverse(std::move(step_inputs));
+    auto rev_result = layer_(rev_step_inputs, input_hidden.second, params.second);
+    std::reverse(rev_result.outputs.begin(), rev_result.outputs.end());
+    auto rev_output = at::stack(rev_result.outputs, 0);
+
+    return {at::cat({fw_output, rev_output}, fw_output.dim() - 1),
+            std::make_pair(fw_result.final_hidden, rev_result.final_hidden)};
+  }
+
+  std::vector<Tensor> reverse(std::vector<Tensor>&& x) const {
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: nit: I wouldn't call a destructive reversing function simply `reverse`

File: aten/src/ATen/native/RNN.cpp [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+
+    // TODO: use split with sizes?
+    auto hidden = input_hidden;
+    for (int64_t i = 0; i < num_steps; ++i) {
+      int64_t batch_size = batch_sizes[i];
+      auto step_input = input.data.narrow(0, input_offset, batch_size);
+      input_offset += batch_size;
+
+      int64_t dec = last_batch_size - batch_size;
+      if (dec > 0) {
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: for when i == 0, I imagine

Comment (ezyang) [reply aten/src/ATen/native/RNN.cpp]: ...and other cases too

File: aten/src/ATen/native/RNN.cpp [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+  output_type operator()(const PackedSequence& input, const hidden_type& input_hidden, const CellParams& params) const override {
+    std::vector<at::Tensor> step_outputs;
+    std::vector<hidden_type> hiddens;
+    int64_t input_offset = 0;
+    int64_t num_steps = input.batch_sizes.size(0);
+    int64_t* batch_sizes = input.batch_sizes.data<int64_t>();
+    int64_t last_batch_size = batch_sizes[0];
+
+    // TODO: use split with sizes?
+    auto hidden = input_hidden;
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: So... do you know what this code is doing? Would you mind adding a comment explaining it? It is breaking my brain.

File: aten/src/ATen/native/RNN.cpp [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+////////////////////////////////////////////////////////////////////////////////
+// apply_layer_stack
+//
+// layers are convenient, but in reality we often want to stack them. this little
+// helper manages slicing of all inputs and parameters, and repeatedly feeds them
+// into the given layer. returns the last layer's outputs, and a vector of final
+// hidden states produced at each level.
+
+Tensor dropout(const Tensor& input, double p) {
+  return at::dropout(input, p, /*train=*/true);
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: Err... is it really always `train=true` for dropout? I see:

```
def StackedRNN(inners, num_layers, lstm=False, dropout=0, train=True):
```

Comment (ezyang) [reply aten/src/ATen/native/RNN.cpp]: Oh I see, you don't bother calling dropout if it is not `train`

File: aten/src/ATen/native/RNN.cpp [hunk 11]
```diff
@@ -0,0 +1,700 @@
...
+  // It's much more useful for us to work on lists of pairs of hx and cx for each layer, so we need
+  // to transpose a pair of those tensors.
+  auto layer_hx = hx.unbind(0);
+  auto layer_cx = cx.unbind(0);
+  int64_t total_layers = layer_hx.size();
+  std::vector<LSTMCell::hidden_type> hiddens;
+  hiddens.reserve(total_layers);
+  for (int64_t i = 0; i < total_layers; ++i) {
+    hiddens.emplace_back(std::move(layer_hx[i]), std::move(layer_cx[i]));
+  }
```
Comment (ezyang) [inline aten/src/ATen/native/RNN.cpp]: ...didn't you already implement this in `pair_vec` and `unpair_vec`?

**Review by ezyang — APPROVED** [review]
Cool stuff.

File: aten/src/ATen/native/RNN.cpp (hunk 8 above)
Comment (apaszke) [reply aten/src/ATen/native/RNN.cpp]: Sorry is this comment meant to be connected with another one? I don't understand what you mean

File: aten/src/ATen/native/RNN.cpp (hunk 7 above)
Comment (apaszke) [reply aten/src/ATen/native/RNN.cpp]: It explicitly takes an rvalue, so the type system should prevent us from destroying non-temporary values.

File: aten/src/ATen/native/RNN.cpp (hunk 6 above)
Comment (apaszke) [reply aten/src/ATen/native/RNN.cpp]: I don't think there's any standard nomenclature... I can change the name if you have better ideas. Essentially it's a non-packed layer, but that sounds awful 😄

File: aten/src/ATen/native/RNN.cpp [hunk 12]
```diff
@@ -0,0 +1,691 @@
...
+
+    auto ingate = chunked_gates[0].sigmoid();
+    auto forgetgate = chunked_gates[1].sigmoid();
+    auto cellgate = chunked_gates[2].tanh();
+    auto outgate = chunked_gates[3].sigmoid();
+
+    auto cy = (forgetgate * cx) + (ingate * cellgate);
+    auto hy = outgate * cy.tanh();
+
+    return std::make_tuple(hy, cy);
```
[the remark this answers is not in the document]
Comment (apaszke) [reply aten/src/ATen/native/RNN.cpp]: Lol what did you use before?

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/RNN.cpp]: the constructor... which means I had to feed it all the template arguments lol
